projects/RabbitMQ.Client/client/api/ConnectionFactory.cs

```diff
@@ -721,7 +724,7 @@ private List<AmqpTcpEndpoint> LocalEndpoints()
             return new List<AmqpTcpEndpoint> { Endpoint };
         }
 
-        private static string EnsureClientProvidedNameLength(string clientProvidedName)
+        private static string? EnsureClientProvidedNameLength(string? clientProvidedName)
```

Comment:
it is possible to add
```c#
[return: NotNullIfNotNull(nameof(clientProvidedName))]
```

Reply:
Would be possible but since this is private, I don't see a big benefit in doing so is very limited as none of the usages would benefit from it. Up to @lukebakken.

Reply:
I agree there isn't much benefit but I do appreciate the additional "documentation" in the code.  I just added it.

Reply:
The C# compiler performs static analysis on the code annotated with NRT.

It's not about being public or private code.

That helps in avoiding nullability tests in private code (you should never trust third party inputs).

```csharp
private static string EnsureClientProvidedNameLength(string clientProvidedName)
```

states that `clientProvidedName` is not null (asserted by the developer that wrote the code), which makes the nullability test redundant.

On the other hand, the method may return a `null` reference, which contradicts the method signature.

If you are adding NRT (and you should), you should honor them. Otherwise, you'll end up with incorrect and unverifiable code.

Reply:
I understand the concept behind, but the NotNullIfNotNull attribute only enhances the definition by telling the compiler the return will not be null if you know the input isn't. 

As this is a private method, it's only called within the same class, where all usages of it are nullable strings as input. So this attribute currently gives no benefit. Hence my comment.

But as @lukebakken said, it can be nice to "document this" / future proof it, which was done now.

Reply:
"if you know" --> "if the compiler can statically prove" 😄